### PR TITLE
MODINREACH-87

### PR DESCRIPTION
**Pull request by andrew-khodyka**
## Purpose
Handle Item Shipped Request for INN-Reach Patron Hold

## Approach

- Add circulation processor for PatronHold
- Add ITEM_SHIPPED CirculationOperation 
- Update inventory client

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/innreach/client/InventoryClient.java [hunk 1]
```diff
@@ -22,4 +24,10 @@
 
   @GetMapping("/items?query=hrid=={hrId}")
   ResultList<InventoryItemDTO> getItemsByHrId(@PathVariable("hrId") String hrId);
+
+  @GetMapping("/items?query=barcode=={barcode}")
+  InventoryItemDTO getItemByBarcode(@PathVariable("barcode") String barcode);
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/client/InventoryClient.java]: The result of query is an array (most probably with single item). It should be wrapped to `ResultList`

File: src/main/java/org/folio/innreach/client/InventoryClient.java [hunk 2]
```diff
@@ -22,4 +24,10 @@
...
+
+  @PutMapping("/items/{itemId}")
+  void updateItem(@PathVariable("itemId") UUID itemId, @RequestBody InventoryItemDTO inventoryItem);
```
Comment (oliinyko) [inline src/main/java/org/folio/innreach/client/InventoryClient.java]: I would suggest to use inventory-storage API instead. Inventory and inventory storage records have a bit different structure and I think it's better to keep using one API as long as it is reasonable. Right now we use inventory storage API for initial record contribution and for some of other logic and it looks much more suitable from backend perspective as it only loads identifiers of reference data without their details.

File: src/main/java/org/folio/innreach/domain/processor/ItemShippedInnReachCirculationProcessor.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  @Override
+  @Transactional
+  public InnReachResponseDTO process(String trackingId, String centralCode, TransactionHoldDTO transactionHold) {
+    var innReachTransaction = transactionRepository.findByTrackingIdAndCentralServerCode(trackingId, centralCode)
+        .orElseThrow(() -> new EntityNotFoundException(String.format(
+            "InnReach transaction with trackingId [%s] and centralCode [%s] not found", trackingId, centralCode)));
+
+    var itemBarcode = transactionHold.getShippedItemBarcode();
```
Comment (dmtkachenko) [inline src/main/java/org/folio/innreach/domain/processor/ItemShippedInnReachCirculationProcessor.java]: image
the request has "itemBarcode" property, from the other side TransactionHoldDTO contains field (and JsonProperty) named "shippedItemBarcode"..
How is the property from json request mapped to the class property?

File: src/main/java/org/folio/innreach/domain/processor/ItemShippedInnReachCirculationProcessor.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  private final InventoryClient inventoryClient;
+
+  @Override
+  public boolean canProcess(String circulationOperationName) {
+    return CirculationOperation.ITEM_SHIPPED.getOperationName().equals(circulationOperationName);
+  }
+
+  @Override
+  @Transactional
+  public InnReachResponseDTO process(String trackingId, String centralCode, TransactionHoldDTO transactionHold) {
```
Comment (dmtkachenko) [inline src/main/java/org/folio/innreach/domain/processor/ItemShippedInnReachCirculationProcessor.java]: According to spec "itemshipped" request can contain only 2 fields:
- callNumber
- itemBarcode

TransactionHoldDTO, according to its schema, has several required fields:
`"required": [
    "transactionTime",
    "pickupLocation",
    "patronId",
    "patronAgencyCode",
    "itemAgencyCode",
    "itemId"
  ]`
Non of these fields are present in the request. How is it expected to make conversion from the request into the object if the mandatory fields are absent?

File: descriptors/ModuleDescriptor-template.json [hunk 5]
```diff
@@ -326,6 +326,11 @@
             "manualblocks.collection.get"
           ]
         },
+        {
+          "methods": ["POST"],
+          "pathPattern": "/inn-reach/d2ir/circ/{circulationOperationName}/{trackingId}/{centralCode}",
+          "permissionsRequired": ["inn-reach.d2ir.circulation.item.post"]
+        },
```
Comment (oliinyko) [inline descriptors/ModuleDescriptor-template.json:333]: Item Shipped endpoint should use PUT method according to API documentation

File: src/main/java/org/folio/innreach/domain/processor/ItemShippedInnReachCirculationProcessor.java (hunk 4 above)
Comment (andrew-khodyka) [reply src/main/java/org/folio/innreach/domain/processor/ItemShippedInnReachCirculationProcessor.java]: @dmtkachenko yes, my fault. I suppose we could use a single DTO to handle all types of request requests, in this case we need to remove all fields from the "required" block to make them "optional".

File: src/main/java/org/folio/innreach/client/InventoryClient.java (hunk 2 above)
Comment (andrew-khodyka) [reply src/main/java/org/folio/innreach/client/InventoryClient.java]: @oliinyko thanks, I'll check

File: src/main/java/org/folio/innreach/domain/processor/ItemShippedInnReachCirculationProcessor.java (hunk 4 above)
Comment (oliinyko) [reply src/main/java/org/folio/innreach/domain/processor/ItemShippedInnReachCirculationProcessor.java]: we could use generic dto `CirculationRequestDTO` without required fields and then map it to more specific DTO  (i.e. `TransactionHoldDTO`) where needed applying validation in runtime with `javax.validation.Validator`

**Review by oliinyko — APPROVED** [review]

File: src/main/java/org/folio/innreach/domain/processor/ItemShippedInnReachCirculationProcessor.java (hunk 3 above)
Comment (andrew-khodyka) [reply src/main/java/org/folio/innreach/domain/processor/ItemShippedInnReachCirculationProcessor.java]: @dmtkachenko, I've changed the payloa, now it's CircualtionRequestDTO that contains the itemBarcode

**Pull request merged**